rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -0,0 +1,155 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.datasource;
+
+import co.rsk.util.MaxSizeHashMap;
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.util.ByteUtil;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class DataSourceWithCache implements KeyValueDataSource {
+    private final KeyValueDataSource base;
+    private final Map<ByteArrayWrapper, byte[]> uncommittedCache;
+    private final Map<ByteArrayWrapper, byte[]> committedCache;
+    private final Set<ByteArrayWrapper> uncommittedKeysToRemove;
+
+    public DataSourceWithCache(KeyValueDataSource base, int cacheSize) {
+        this.base = base;
+        this.uncommittedCache = new HashMap<>();
+        this.committedCache = new MaxSizeHashMap<>(cacheSize, true);
+        this.uncommittedKeysToRemove = new HashSet<>();
+    }
+
+    @Override
+    public byte[] get(byte[] key) {
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        if (uncommittedKeysToRemove.contains(wrappedKey)) {
+            return null;
+        }
+
+        if (committedCache.containsKey(wrappedKey)) {
+            return committedCache.get(wrappedKey);
+        }
+        if (uncommittedCache.containsKey(wrappedKey)) {
+            return uncommittedCache.get(wrappedKey);
+        }
+
+        byte[] value = base.get(key);
+        //null value, as expected, is allowed here to be stored in committedCache
+        committedCache.put(wrappedKey, value);
+        return value;
+    }
+
+    @Override
+    public byte[] put(byte[] key, byte[] value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        // here I could check for equal data or just move to the uncommittedCache.
+        byte[] priorValue = committedCache.get(wrappedKey);
+        if (priorValue != null && Arrays.equals(priorValue, value)) {
+            return value;
+        }
+
+        committedCache.remove(wrappedKey);
+        uncommittedCache.put(wrappedKey, value);
+        uncommittedKeysToRemove.remove(wrappedKey);
+        return value;
+    }
+
+    @Override
+    public void delete(byte[] key) {
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+        // fully delete this element
+        committedCache.remove(wrappedKey);
```

Comment:
Minor minor: prevention of deleted deletes

Reply:
do you mean to add a `committedCache.containsKey(wrappedKey)` check before invoking the removal itself? I don't see what would be the benefit

Reply:
it prevents the  uncommittedCache.put(wrappedKey, null) to perform a later base.delete(), because we assume the base DB is slower than a committedCache.containsKey()